[Sema] Adapt force optional fix to handle multiple levels of optional

LucianoPAlmeida opened the pull request:
<!-- What's in this pull request? -->
Adapt force optional fix to handle multiple levels of optional changing signature to take only optional type and unwrap count(instead of from and to types) which make sense since the force optional is to unwrap an optional to produce the unwrapped type. So let me know what you think =]

<!--
If this pull request resolves any GitHub issues, link them.
For information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

LucianoPAlmeida commented:
@swift-ci Please test

xedin commented on test/Constraints/optional.swift:
```diff
@@ -574,15 +574,14 @@ do {
-    // expected-note@-2 {{coalesce using '??' to provide a default when the optional value contains 'nil'}}
-    // expected-note@-3 {{force-unwrap using '!' to abort execution if the optional value contains 'nil'}}
+    // expected-error@-1 {{value of optional type '(any P)??' must be unwrapped to a value of type 'any P'}}
+    // expected-note@-2 {{coalesce using '??' to provide a default when the optional value contains 'nil'}} {{17-17= ?? <#default value#>}}
```
This is the problematic part - suggesting a single `??` is incorrect in this case, it has to be `(... ?? ...) ?? ...`...

LucianoPAlmeida replied:
I have changed to not emit coalesce fix-it for more than one level. More than that looks like a too complicated chain of defaults to suggest. Let me know what you think :)

xedin commented on lib/AST/Type.cpp:
```diff
@@ -906,9 +906,13 @@ Type TypeBase::lookThroughAllOptionalTypes(SmallVectorImpl<Type> &optionals){
+    type = objType;
+    ++depth;
+  }
+  return depth;
```
I think we should keep this the way it was really or change it in a separate commit.

LucianoPAlmeida replied:
This was in a different commit than actual changes already, but if you think is worth another NFC PR for that I can remove and open separately.

xedin replied:
Let's land this separately because it's not technically necessary for the fix since the resulting count is the same.

LucianoPAlmeida replied:
Sure, I'll land it separately =]

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -89,13 +89,21 @@ ForceDowncast *ForceDowncast::create(ConstraintSystem &cs, Type fromType,
 bool ForceOptional::diagnose(const Solution &solution, bool asNote) const {
   MissingOptionalUnwrapFailure failure(solution, getFromType(), getToType(),
-                                       getLocator());
+                                       UnwrapCount, getLocator());
```
So this is where the fun begins, see my comment in the test case :) I think we need to "stack" the failures while diagnosing a force-unwrap fix because it requires `??` per level of optional unwrapped.

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -857,7 +860,8 @@ class ForceOptional final : public ContextualMismatch {
 
-  static ForceOptional *create(ConstraintSystem &cs, Type fromType, Type toType,
+  static ForceOptional *create(ConstraintSystem &cs, Type fromType,
+                               unsigned unwrapCount,
```
I think it would be safer to compute this in initializer - we can take optionality of `from` and `to` types, diff them on construction of the fix and add an accessor method to retrieve the number of unwraps so it could be used in `recordFix`. Also note that we'd want to reserve `toType` provided by the caller because it's not necessarily equal to unwrapped from since there could be implicit conversions involved.

LucianoPAlmeida commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -12972,16 +12972,11 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
-    SmallVector<Type, 4> unwraps2;
-    type2->lookThroughAllOptionalTypes(unwraps2);
-
-    auto impact = unwraps1.size() != unwraps2.size()
```
Still have to think about this a bit more, in one specific case where fromObjectType is a type variable we still record force optional fix with from and to with same level of optional if object types match which means toObjectType is also a type variable probably. Have to look a bit more

xedin replied:
Can we "delay" a fix here by returning "unsolved" until the type variable that represents object type is resolved?

LucianoPAlmeida replied:
This makes sense, I'll have more time over the weekend to work on that. Let's try that and see :)

LucianoPAlmeida replied:
I just looks like we somehow already "delaying" but only for `Equal` constraints https://github.com/apple/swift/blob/99dcabd1a62475059dea61f582ab180f80db67b0/lib/Sema/CSSimplify.cpp#L4456
not in the sense that we return "unsolved" and put the constraint back to wait until a fixed type is bound, but in that force unwrap is not recorded if fixed types aren't bound for equality constraints only. To me is still a bit confusing why we allowed equal number of unwraps record a force optional fix when type variables are involved because it could or not be an optional so force optional does not look correct in this case. But is possible I'm missing something.

xedin replied:
Yeah, that doesn't make much sense to report a fix when there is no unwrapping needed seems like that should be an assert instead.

xedin replied:
We need to figure out in what situations that happens (if any) and address them at the source.

LucianoPAlmeida replied:
> Yeah, that doesn't make much sense to report a fix when there is no unwrapping needed seems like that should be an assert instead.

That is actually how this was caught because `ForceOptional` now has an assert on optional count diff so that is why I removed the type variable check from https://github.com/apple/swift/blob/3c4dd9fad48a0843b43a444a77e18bc65d9e8e54/lib/Sema/CSSimplify.cpp#L4571

So that is why I was asking about why we only exit early for Equal constraints above, but allowing typevars otherwise which potentially ends up in a equal optionality recorded as ForceOptional. 

> We need to figure out in what situations that happens (if any) and address them at the source.

Sure, I'll try go back to this, was a bit confusing. Thanks for the input =]

xedin replied:
> So that is why I was asking about why we only exit early for Equal constraints above

I think the reason there is partially to optional evaluation expressions which preserve optionality, try extending it and see what fails.

LucianoPAlmeida replied:
Sorry for taking a while to respond 😅 

>  try extending it and see what fails.

Removing the `Equal` match kind from the check only breaks on one case in the whole test suite involving a dependent member with optional base https://godbolt.org/z/jv48h8W63. But in this case makes sense to record the force optional fix because we have a `$T0?` bind `Int`.

xedin replied:
No worries! I guess we could narrow the check to allow it down to that situation only then?

LucianoPAlmeida replied:
I'm thinking a bit about this, this condition seems fine we for sure could narrow it but that makes sense the way it is.
The problematic case I think is this `!fromObjectType->is<TypeVariableType>()`
https://github.com/apple/swift/blob/6676fbfd19781dd6ed1d59bb84146aff336c24d4/lib/Sema/CSSimplify.cpp#L4571
I don't think we should allow a force optional fix when fromObjType is a type variable with less or equal unwraps then to. Because we can't know for sure it would be an optional(as previous comment said "it may be" but it cannot). I already removed that and it seems not to break any diagnostic in the suit and neither the new assert in ForceOptional ` assert(fromOptionals > toOptionals)` so `getUnwrapCount()` will always be > 0. That makes me think that it should be ok. Do you think that makes sense? 
I'll run the full suit of tests in CI just to make sure.

xedin replied:
> I don't think we should allow a force optional fix when fromObjType is a type variable with less or equal unwraps then to. 

Yes, that is correct. The check you mentioned is checking exactly that, it only allows the fix if "from" is not a type variable, that the situation when we know for sure that there is a missing unwrap.

LucianoPAlmeida replied:
You mean this one https://github.com/apple/swift/blob/6676fbfd19781dd6ed1d59bb84146aff336c24d4/lib/Sema/CSSimplify.cpp#L4571
? If I understand correctly this is the other way around, like only exit early in case from less or equal and from is not a type var which is what is allowing type vars with equal number of unwraps being recorded as force optional. Let me take a look again to make sure, is possible I miss understood the condition

LucianoPAlmeida replied:
Looks like it's indeed the other way around, and that is what was allowing type vars to ignore this condition `fromUnwraps <= toUnwraps`.

xedin replied:
You are right, it should be the other way around - the comment states one thing but the code does the other :)

LucianoPAlmeida replied:
> No worries! I guess we could narrow the check to allow it down to that situation only then?

Just did that adjustment, just don't record force optional if type variables are involved expect for that case. Let me know what you think.

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -843,17 +843,26 @@ class MarkGlobalActorFunction final : public ContextualMismatch {
                            locator) {
     assert(fromType && "Base type must not be null");
     assert(fromType->getOptionalObjectType() &&
            "Unwrapped type must not be null");
```
A better message here would be - `from type must be optional`

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -843,17 +843,26 @@ class MarkGlobalActorFunction final : public ContextualMismatch {
 
 public:
   std::string getName() const override { return "force optional"; }
+  unsigned int getUnwrapCount() const { return UnwrapCount; }
```
Nit: just `unsigned` is fine since we store `unsigned` or let's store `unsigned int` too.

LucianoPAlmeida replied:
Fixed!

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -1411,6 +1411,9 @@ void MissingOptionalUnwrapFailure::offerDefaultValueUnwrapFixIt(
     DeclContext *DC, const Expr *expr) const {
   assert(expr);
 
+  if (UnwrapCount != 1) 
```
Nit: Needs a comment explaining why this is so.

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -796,10 +796,24 @@ class DroppedGlobalActorFunctionAttr final : public ContextualFailure {
 /// which require some type of force-unwrap e.g. "!" or "try!".
 class MissingOptionalUnwrapFailure final : public ContextualFailure {
+  unsigned UnwrapCount;
+  Type UnwrappedType;
```
This could be computed when needed right? Probably no need to store it.

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -809,8 +823,17 @@ class MissingOptionalUnwrapFailure final : public ContextualFailure {
+  std::string getForceUnwrapString() const {
+    std::string forceUnwrapStr;
+    unsigned i = 0;
+    while (i != UnwrapCount) {
```
This loop is no-op now, isn't it? The diagnostic returns if number of unwraps is > 1.

LucianoPAlmeida replied:
This is for force unwrap note, the diagnostic note we skip is coalesce one

LucianoPAlmeida committed:
[Sema] Adapt ForceOptional fix to take unwrap count as param and handle more than one level of optionality

LucianoPAlmeida commented:
@swift-ci Please test MacOS Platform

LucianoPAlmeida commented:
@swift-ci Please test MacOS Platform

LucianoPAlmeida commented:
@swift-ci Please test MacOS Platform

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4552,36 +4552,28 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
+  // we fail but fromUnwraps <= toUnwraps and fromObjectType can be bound to
+  // this concrete type, we know a force optional is the correct fix.
+  if (!isBindingToFixedType && (fromObjectType->is<TypeVariableType>() ||
+                                toObjectType->is<TypeVariableType>())) {
```
So how about this idea - let's remove `isBindingToFixedType` and simplify this check into:

```
if (fromObjectType->is<TypeVariableType>() || toObjectType->is<TypeVariableType>())
```

and at the same time expand return type of this method from simple true/false into a `TypeMatchResult` - in case when this check triggers it should return `amibiguous` and all the call sites in `repairFailures` would `break` if it does, this way we'd make sure that there are no fixes would be recorded in this case which I believe should also fix the case with `?.` syntax you mentioned.

LucianoPAlmeida replied:
Right, sounds good. I'll take a look on that approach in more detail soon. Initially with a quick change that cause that case specifically failed and end up without a fix and default diagnostic, but that may be just me messing something up. Let's see... thank you for the idea!

LucianoPAlmeida replied:
Unfortunately I couldn't make it work, trying delay that way for this case in specific only end leaving the expression without a diagnostic because that would never comeback to that failure a re-attempt repair with fixed types assigned.

xedin committed:
[CSSimplify] `repairViaOptionalUnwrap` to return `TypeMatchResult`

xedin committed:
[CSSimplify] Allow `repairFailures` indicate that it is confused

Sometimes it's not possible to determine what fix to use if
involved types are not resolved enough. `repairFailures` needs
a way to indicate to `matchTypes` that the matching has to be
delayed until the solver has a change to infer more type
information.

xedin committed:
[CSSimplify] Allow `repairViaOptionalUnwrap` to produce ambiguous result if one of the object types is a type variable

LucianoPAlmeida committed:
[Sema] Update missing new code
